File: prog1/Lex.java
import java.io.*;
import java.util.Scanner;
   
class Lex{
   public static void main(String[] args) throws IOException{
      Scanner in = null;
      PrintWriter out = null;
      String line = null;
      int i, n, lineNumber = 0;
      
      
      //reports to stderr if more or less than 2 arguments
      if(args.length != 2){
         System.err.println("Incorrect Number of arguments");
         System.exit(1);
      }
      
      //initialized the in and out arguments
      in = new Scanner(new File(args[0]));
      out = new PrintWriter(new FileWriter(args[1]));

      int test = countLineNumber(in);
      String[] lineArray = new String[test];
      
      //inserts the strings from the file into the string array
      Scanner in2 = new Scanner(new File(args[0]));
      while(in2.hasNextLine()){
      for(int a = 0;a < test;a++){
	  lineArray[a] = in2.nextLine();
        }
      }

      //call lexSort which sorts the indices according to array
      List lex = lexSort(lineArray);

      //prints the values of lexSort indices as they are in lineArray
      for(lex.moveTo(0); lex.getIndex() >= 0; lex.moveNext()){
	   out.println(lineArray[lex.getElement()]);  
      }

      in.close();
      out.close();
   }

    //function to count the number of lines in the file
    public static int countLineNumber(Scanner in){
       int lineNumber = 0;
       while(in.hasNextLine()){
	   lineNumber++;
	   in.nextLine();
       }
       return lineNumber;
   }

   public static List lexSort(String [] lineArray){
       List lex = new List();
       //append 0 to the front of empty list
       if(lex.length() == 0) lex.append(0);

       //follows the pseudo-code for insertionSort
       for(int i = 1; i < lineArray.length; i++){
	   //temporary string that contains the array index being compared to the cursor string index
	   String tmp = lineArray[i];
           int j = i - 1;
	   lex.moveTo(j);
	   //if the tmp string compared to the element of the cursor is less then it comes before. 
           while(j > -1 && tmp.compareTo(lineArray[lex.getElement()]) < 0){
	           lex.movePrev();
		   j--;
		   }
	   //if cursor falls off then add to front
	   if(lex.getIndex() == -1){
	       lex.prepend(i);
	       //if not then insert after the cursor location.
	   } else {
	       lex.insertAfter(i);
	   }
	 } return lex;
  }
} 

File: prog2/Lex.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>

#include "List.h"

#define MAX_LEN 160

//method for the strdup
char *strdup(const char *a){
    char *dup = malloc(strlen(a) + 1);
    if (dup == NULL) return NULL;
    strcpy(dup, a);
    return dup;
}

List lexSort(char** lineArray, int length){
       List lex = newList();
       //append 0 to the front of empty list
       if(length > 0) append(lex,0);
       //follows the pseudo-code for insertionSort
       for(int i = 1; i < length; i++){
	   //temporary string that contains the array index being compared to the cursor string index
	   char* tmp = lineArray[i];
           int j = i - 1;
	   moveTo(lex,j);
	   //if the tmp string compared to the element of the cursor is less then it comes before. 
           while(j > -1 && (strcmp(tmp,lineArray[getElement(lex)]) < 1)){
	           movePrev(lex);
		   j--;
		   }
	   //if cursor falls off then add to front
	   if(getIndex(lex) == -1){
	       prepend(lex,i);
	       //if not then insert after the cursor location.
	   } else {
	       insertAfter(lex,i);
	   }
	 } return lex;
  }

int main(int argc, char * argv[]){

   int count=0;
   FILE *in, *out;
   char line[MAX_LEN];

   // check command line for correct number of arguments
   if( argc != 3 ){
      printf("Usage: %s <input file> <output file>\n", argv[0]);
      exit(1);
   }

   // open files for reading and writing 
   in = fopen(argv[1], "r");
   out = fopen(argv[2], "w");
   if( in==NULL ){
      printf("Unable to open file %s for reading\n", argv[1]);
      exit(1);
   }
   if( out == NULL ){
      printf("Unable to open file %s for writing\n", argv[2]);
      exit(1);
   }
   
   //loops through the file and counts the strings
   while(fgets(line, MAX_LEN, in) != NULL)  {
      count++;
   }

   //closes the counter scanner on the file in
   fclose(in);
   FILE *in2 = fopen(argv[1], "r");
 
   //creates a new string array called lineArray
   char** lineArray = malloc(sizeof(char**) * count);

   //for loop that 
   for(int i = 0; i < count; i++){
       fgets(line, MAX_LEN, in2);
       lineArray[i] = strdup(line);
   } 

   //references the lexSort function that alphabetizes the indecies
   List lex = lexSort(lineArray, count); 

   //print the sorted List's elements
   for(moveTo(lex,0); getIndex(lex) >= 0; moveNext(lex)){
        fprintf(out, "%s", lineArray[getElement(lex)]);
   }

   //frees the array and list
   free(lineArray); 
   freeList(&lex);

   /* close files */
   fclose(out);
   fclose(in2);

   return(0);
}

